Add redux variant to standalone variant migration guide (#1653)
eclipsesource/jsonforms: MIGRATION.md

  {
    jsonforms: {
      cells: materialCells,
      renderers: materialRenderers
    }
  }
);
store.dispatch(Actions.init(data, schema, uischema));
return (
  <Provider store={store}>
    <JsonFormsReduxContext>
      <JsonFormsDispatch />
    </JsonFormsReduxContext>
  </Provider>
);
```

Instead of creating a store and passing the required information to that store, we rather pass it directly to the `<JsonForms .../>` component:
```ts
return (
  <JsonForms
    schema={schema}
    uischema={uischema}
    data={data}
    renderers={materialRenderers}
    cells={materialCells}
  />
);
```

### Example 2: Register a custom renderer
Another commonly used action is the 'register renderer' action.

With Redux this could look like this:
```ts
store.dispatch(Actions.registerRenderer(customControlTester, CustomControl));
```

Within the standalone version, the renderer can just be provided to the `<JsonForms .../>` element like this:
```ts
const renderers = [
  ...materialRenderers,
  // register custom renderer
  { tester: customControlTester, renderer: CustomControl }
];

const MyApp = () => (
  <JsonForms
    // other necessary declarations go here...
    renderers={renderers}
  />

> );

Reviewer: ```suggestion
);

```

> ```

Reviewer: ```suggestion

```
```

### Example 3: Listen to data and validation changes
The `JsonForms` component offers to register a listener which is notified whenever `data` and `errors` changes:

```ts
const MyApp = () => {
  const [data, setData] = useState();
  return (
    <JsonForms
      data={data}
      // other necessary declarations go here...
      onChange={({ data, errors }) => setData(data)}
    />
  );
};

> ## Case 2: Keep using Redux variant of JSON Forms with our fallback (not recommend)

Reviewer: ```suggestion
```
## Case 2: Keep using Redux variant of JSON Forms with our fallback (not recommend)
```

# Migrating to JSON Forms 2.5 for React users

In version 2.5 we made the `redux` dependency within the `react` package optional.
Users of the JSON Forms React standalone version (i.e. without Redux) don't need to change anything.
In contrary you no longer need to install 'redux' and 'react-redux' to use JSON Forms.

Users of the JSON Forms Redux variant need to perform some changes.

Basically there are two different approaches: 
1. Migrate your app to the standalone variant
2. Keep using the Redux variant of JSON Forms

Below you can find some guidance about each approach.

> ## Case 1: Migrate to the standalone variant (recommend)

Reviewer: ```suggestion
## Case 1: Migrate to the standalone variant (recommended)
```

The standalone JSON Forms variant is the new default and the main focus for new features and bug fixes.
We definitely recommend migrating to this version as soon as possible.

> All current redux functionally can also be achieved with the standalone version.

Reviewer: ```suggestion
All current Redux functionally can also be achieved with the standalone version.
```

### Example 1: Init action
Previously the store was initialized like this:
```ts
const store = createStore(
  combineReducers({ jsonforms: jsonformsReducer() }),
  {
    jsonforms: {
      cells: materialCells,
      renderers: materialRenderers
    }
  }
);
store.dispatch(Actions.init(data, schema, uischema));
return (
  <Provider store={store}>
    <JsonFormsReduxContext>
      <JsonFormsDispatch />
    </JsonFormsReduxContext>
  </Provider>
);
```

Instead of creating a store and passing the required information to that store, we rather pass it directly to the `<JsonForms .../>` component:
```ts
return (
  <JsonForms
    schema={schema}
    uischema={uischema}
    data={data}
    renderers={materialRenderers}
    cells={materialCells}
  />
);
```

### Example 2: Register a custom renderer

> Another common used action is the renderer register action.

Reviewer: ```suggestion
Another commonly used action is the 'register renderer' action.
```

With Redux this could look like this:
```ts
store.dispatch(Actions.registerRenderer(customControlTester, CustomControl));
```

> Within the standalone version, the renderer would just be provided to the `<JsonForms .../>` element like this:

Reviewer: ```suggestion
Within the standalone version, the renderer can just be provided to the `<JsonForms .../>` element like this:
```

```ts
const renderers = [
  ...materialRenderers,
  // register custom renderer
  { tester: customControlTester, renderer: CustomControl }
];

const MyApp = () => (
  <JsonForms
    // other necessary declarations go here...
    renderers={renderers}
  />
);
```

> 

Reviewer: ```suggestion
### Example 3: Listen to data and validation changes
The `JsonForms` component offers to register a listener which is notified whenever `data` and `errors` changes:

```ts
const MyApp = () => {
  const [data, setData] = useState();
  return (
    <JsonForms
      data={data}
      // other necessary declarations go here...
      onChange={({ data, errors }) => setData(data)}
    />
  );
};
```
The event will also be fired during the first rendering for any potential data changes (via `AJV` default support) as well as the initial validation result.
```
